kernel/src/common/regs/README.md

```diff
@@ -0,0 +1,207 @@
+# Register/Bitfield Interface for Tock
+
+This module provides an interface for defining and manipulating memory mapped
+registers and bitfields.
+
+## Defining registers
+
+The module provides three types for working with memory mapped registers: `ReadWrite`,
+`ReadOnly`, and `WriteOnly`, providing read-write, read-only, and write-only functionality,
+respectively.
+
+Defining the registers is similar to the C-style approach, where each register
+is a field in a packed struct:
+
+```rust
+use common::regs::{ReadOnly, ReadWrite, WriteOnly};
+
+#[repr(C, packed)]
+struct Registers {
+    // Control register: read-write
+    // The 'Control' parameter constrains this register to only use fields from
+    // a certain group (defined below in the bitfields section).
+    cr: ReadWrite<u8, Control>,
+
+    // Status register: read-only
+    s: ReadOnly<u8, Status>
+
+    // Registers can be bytes, halfwords, or words:
+    // Note that the second type parameter can be omitted, meaning that there
+    // are no bitfields defined for these registers.
+    byte0: ReadWrite<u8>,
+    byte1: ReadWrite<u8>,
+    short: ReadWrite<u16>,
+    word: ReadWrite<u32>
+
+    // Etc.
+}
+```
+
+## Defining bitfields
+
+Bitfields are defined through the `bitfields!` macro:
+
+```rust
+bitfields! [
+    // First parameter is the register width for the bitfields. Can be u8, u16,
+    // or u32.
+    u8,
+
+    // Each subsequent parameter is a register abbreviation, its descriptive
+    // name, and its associated bitfields.
+    // The descriptive name defines this 'group' of bitfields. Only registers
+    // defined as ReadWrite<_, Control> can use these bitfields.
+    CR Control [
+        // Bitfields are defined as:
+        // name (Mask(mask), shift) [ /* optional named values */ ]
+
+        // This is a two-bit field which includes bits 4 and 5
+        RANGE (Mask(0b11), 4) [
+            // Each of these defines a name for a value that the bitfield can be
+            // written with or matched against. Note that this set is not exclusive--
+            // the field can still be written with arbitrary constants.
+            VeryHigh = 0,
+            High = 1,
+            Low = 2
+        ],
+
+        // A common case is single-bit bitfields, which usually just mean
+        // 'enable' or 'disable' something. This syntax is a shorthand, defining
+        // EN as a field over bit 3, and INT as a field over bit 2.
+        EN 3 [],
+        INT 2 []
+    ],
+
+    // Without the explanatory comments like above, bitfield definition is quite compact:
+    // Status register
+    S Status [
+        TXCOMPLETE 0 [],
+        TXINTERRUPT 1 [],
+        RXCOMPLETE 2 [],
+        RXINTERRUPT 3 [],
+        MODE (Mask(0b11), 4) [
+            FullDuplex = 0,
+            HalfDuplex = 1,
+            Loopback = 2,
+            Disabled = 3
+        ],
+        ERRORCOUNT (Mask(0b11), 6) []
+    ]
+]
+```
+
+
+## Example: Using registers and bitfields
```

Comment:
Maybe before this it would help to just have a listing of the functions. Something like:

```
ReadOnly
.read()  // Get the value of the bitfield.
.is_set()
.matches()

WriteOnly
...

ReadWrite
...
```

Reply:
Added. Not sure if including the type parameters in the listing was confusing or helpful.